### PR TITLE
Allow environments to emit categorical observations

### LINKED ISSUE #277: rollout function is incompatible with onehot_obs Agent?
With an initial setup like so:

```python
from jax import random as jr
from pymdp.agent import Agent
from pymdp.envs import TMaze, rollout

key = jr.PRNGKey(0)
batch_size=1
T = 5
env = TMaze(
    batch_size=batch_size,
)
A, A_dependencies = env.generate_A()
B, B_dependencies = env.generate_B()
agent = Agent(
    A=A,
    B=B,
    A_dependencies=A_dependencies,
    B_dependencies=B_dependencies,
    batch_size=batch_size,
    policy_len=T,
    onehot_obs=True,
)
_, info, _ = rollout(agent, env, num_timesteps=T, rng_key=key)
```
`rollout` raises an exception when trying to call `agent.infer_states(...)`. 

This does not occur when setting `onehot_obs=False` in the `Agent` constructor.  We should either:
1. fix rollout to handle one hot encoded observations (preferable)
2. remove the flag from the agent constructor

if we keep support for one hot encoded observations, we should add unit tests to cover this option

### PR DESCRIPTION
**Pull request by conorheins**
Fixes #277, re-implementation of @Arun-Niranjan's approach in #328 

## Summary
- add a `categorical_obs` option to `PymdpEnv` so envs can emit one-hot observation samples directly
- thread that option through the refactored environment classes and update TMaze rendering to decode categorical traces
- add rollout regression coverage showing categorical env observations match the discrete path semantically and that TMaze rollout works with `categorical_obs=True`

## Testing
- `python -m pytest test/test_rollout_function.py test/test_cue_chaining_env.py test/test_categorical_observations.py test/test_tmaze_envs.py -q`

### CODE REVIEW

File: pymdp/envs/env.py [hunk 1]
```diff
@@ -380,6 +386,17 @@ def _sample_obs(
...
         keys = list(jr.split(key, len(obs_probs)))
-        new_obs = jtu.tree_map(cat_sample, keys, obs_probs)
-        new_obs = jtu.tree_map(lambda x: jnp.expand_dims(x, -1), new_obs)
+        obs_idx = jtu.tree_map(cat_sample, keys, obs_probs)
+
+        if self.categorical_obs:
+            new_obs = [
+                jnp.expand_dims(
+                    nn.one_hot(jnp.asarray(obs_m, dtype=jnp.int32), probs_m.shape[0]),
+                    axis=0,
```
[the remark this answers is not in the document]
Comment (conorheins) [reply pymdp/envs/env.py]: Implemented. Actually `reset()` / `step()` are intended to be unbatched and batching is expected to happen externally via `vmap`, so the previous code was valid for the documented path. But using `probs_m.shape[-1]` and inserting the singleton time axis at -2 is the more robust formulation and preserves the current unbatched behavior while handling any future batched reuse correctly.

File: pymdp/envs/cue_chaining.py [hunk 2]
```diff
@@ -51,6 +51,7 @@ def __init__(
         reward_condition: int | None = None,
         cue2_names: Sequence[str] | None = None,
         reward_condition_names: Sequence[str] | None = None,
+        categorical_obs: bool = False,
     ) -> None:
         """Initialize the cue-chaining environment.
```
[the remark this answers is not in the document]
Comment (conorheins) [reply pymdp/envs/cue_chaining.py:56]: addressed now with 3cb0a6562cedecd7c84ec9dfe7d78fbec4d6b5a3

File: pymdp/envs/grid_world.py [hunk 3]
```diff
@@ -75,6 +75,7 @@ def __init__(
         initial_position: Optional[Tuple[int, int] | int] = None,
         include_stay: bool = True,
         success_prob: float = 1.0,
+        categorical_obs: bool = False,
     ) -> None:
```
[the remark this answers is not in the document]
Comment (conorheins) [reply pymdp/envs/grid_world.py:81]: now addressed with 3cb0a6562cedecd7c84ec9dfe7d78fbec4d6b5a3

File: pymdp/envs/tmaze.py [hunk 4]
```diff
@@ -48,6 +48,7 @@ def __init__(
         cue_validity: float = 0.95,
         reward_condition: Optional[int] = None,
         dependent_outcomes: bool = False,
+        categorical_obs: bool = False,
         *,
```
[the remark this answers is not in the document]
Comment (conorheins) [reply pymdp/envs/tmaze.py:52]: now addressed with 3cb0a6562cedecd7c84ec9dfe7d78fbec4d6b5a3

File: pymdp/envs/tmaze.py [hunk 5]
```diff
@@ -303,7 +332,13 @@ def render(
 
         plt.clf()
 
-       # create n x n subplots for the batch_size
+        def _decode_obs(modality: int, batch_idx: int) -> int:
+            obs_value = observations[modality][batch_idx, 0]
+            if jnp.ndim(obs_value) > 0:
+                return int(jnp.argmax(obs_value))
+            return int(obs_value)
```
[the remark this answers is not in the document]
Comment (conorheins) [reply pymdp/envs/tmaze.py]: Implemented. You’re right that the previous ndim > 0 check was too broad and could mis-handle singleton-shaped discrete observations by routing them through argmax. I changed the decoder to use argmax only when the observation slice has more than one element, and otherwise squeeze() to a scalar first. I also added a TMaze render regression test covering both singleton discrete inputs and categorical one-hot inputs.
